Type can no longer be inferred in 1.49

<!--
Thank you for filing a regression report! 🐛 A regression is something that changed between versions of Rust but was not supposed to.

Please provide a short summary of the regression, along with any information you feel is relevant to replicate it.
-->

### Code

I tried this code: [Johni0702/rtp on the dtls-srtp Branch](https://github.com/Johni0702/rtp/tree/6c0223d8a43ea5251574c541e72461b5d18a3906)

I expected to see this happen: Compiles successfully as in 1.48

Instead, this happened: Error E0282 in `src/rfc3711.rs`

```
error[E0282]: type annotations needed
   --> src/rfc3711.rs:477:19
    |
476 |         let iv = iv ^ (BigUint::from(1_u8) << (context.session_encr_key.len() * 8));
    |             -- consider giving `iv` a type
477 |         let iv = &iv.to_bytes_be()[1..context.session_encr_key.len() + 1];
    |                   ^^ cannot infer type
    |
    = note: type must be known at this point

error[E0282]: type annotations needed
   --> src/rfc3711.rs:511:19
    |
510 |         let iv = iv ^ (BigUint::from(1_u8) << (context.session_encr_key.len() * 8));
    |             -- consider giving `iv` a type
511 |         let iv = &iv.to_bytes_be()[1..context.session_encr_key.len() + 1];
    |                   ^^ cannot infer type
    |
    = note: type must be known at this point

error: aborting due to 2 previous errors
```

### Version it worked on

<!--
Provide the most recent version this worked on, for example:

It most recently worked on: Rust 1.48
-->

It most recently worked on: 1.48

### Version with regression

<!--
Provide the version you are using that has the regression.
-->

`rustc --version --verbose`:
```
rustc 1.49.0 (e1884a8e3 2020-12-29)
binary: rustc
commit-hash: e1884a8e3c3e813aada8254edfa120e85bf5ffca
commit-date: 2020-12-29
host: x86_64-unknown-linux-gnu
release: 1.49.0
```

</details>

<!--
If you know when this regression occurred, please add a line like below, replacing `{channel}` with one of stable, beta, or nightly.

@rustbot modify labels: +regression-from-stable-to-{channel}-regression-untriaged
-->


I tried to minimize the code necessary for reproduction and I got it down to:

```rust
//file: src/lib.rs
use num::BigUint; // Cargo.toml: num = "0.1"

pub trait Protocol {
    type PacketIndex: Into<u64> + Into<BigUint>;
}

fn from_nothing<T>() -> T {
    todo!()
}

pub fn decrypt_portion<P: Protocol>(index: P::PacketIndex) {
    let iv: BigUint = from_nothing();
    let len: usize = from_nothing();
    let iv = iv ^ (index.into() << 16);
    let iv = iv ^ (BigUint::from(1_u8) << (len * 8));
    let _iv: &[u8] = &iv.to_bytes_be()[1..len + 1];
}
```

Minimized further:

```rust
use num::BigUint; // Cargo.toml: num = "0.1"

pub trait Protocol: Sized + Default {
    type PacketIndex: Into<u64> + Into<BigUint>;
}

pub fn decrypt_portion<P: Protocol>(index: P::PacketIndex) {
    let iv: BigUint = loop{};
    let len: usize = loop{};
    let iv = iv ^ (index.into() << 16);
    let iv = iv ^ (BigUint::from(1_u8) << (len * 8));
    let _iv: &[u8] = &iv.to_bytes_be()[1..len + 1];
}
```

([Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=4babe100f138922aed611357bccedc37))

Errors:

```
   Compiling playground v0.0.1 (/playground)
warning: unreachable statement
 --> src/lib.rs:9:5
  |
8 |     let iv: BigUint = todo!();
  |                       ------- any code following this expression is unreachable
9 |     let len: usize = todo!();
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^ unreachable statement
  |
  = note: `#[warn(unreachable_code)]` on by default

error[E0282]: type annotations needed
  --> src/lib.rs:12:23
   |
11 |     let iv = iv ^ (BigUint::from(1_u8) << (len * 8));
   |         -- consider giving `iv` a type
12 |     let _iv: &[u8] = &iv.to_bytes_be()[1..len + 1];
   |                       ^^ cannot infer type
   |
   = note: type must be known at this point

error: aborting due to previous error; 1 warning emitted

For more information about this error, try `rustc --explain E0282`.
error: could not compile `playground`

To learn more, run the command again with --verbose.
```

--- a/tests/ui/type-inference/regression-issue-81317.rs
+++ b/tests/ui/type-inference/regression-issue-81317.rs
@@ -1,4 +1,33 @@
 // Regression test for #81317: type can no longer be infered as of 1.49
+//
+// The problem is that the xor operator and the index.into() each have two candidate impls that could apply
+// { S as BitXor<S>, S as BitXor<&'a S> } for xor and
+// { T::I as Into<u64>, T::I as Into<S> } for index.into()
+// previously inference was able to infer that the only valid combination was
+// S as BitXor<S> and T::I as Into<S>
+//
+// after rust-lang/rust#73905 this is no longer infered
+//
+// the error message could be better e.g. when iv is unused or has an an explicitly specified type S
+// there is currently the following help message
+//
+// error[E0284]: type annotations needed
+//   --> src/main.rs:13:24
+//    |
+// 42 |     let iv = S ^ index.into();
+//    |                -       ^^^^
+//    |                |
+//    |                type must be known at this point
+//    |
+//    = note: cannot satisfy `<S as BitXor<_>>::Output == _`
+// help: try using a fully qualified path to specify the expected types
+//    |
+// 42 -     let iv = S ^ index.into();
+// 42 +     let iv = S ^ <<T as P>::I as Into<T>>::into(index);
+//
+// this is better as it's actually sufficent to fix the problem,
+// while just specifying the type of iv as currently suggested is insufficent
+//
 //@ check-fail
 
 use std::ops::BitXor;

--- a/tests/ui/type-inference/regression-issue-81317.rs
+++ b/tests/ui/type-inference/regression-issue-81317.rs
@@ -1,0 +1,40 @@
+// Regression test for #81317: type can no longer be infered as of 1.49
+//@ check-fail
+
+use std::ops::BitXor;
+
+pub struct S;
+
+pub trait P {
+    type I: Into<u64> + Into<S>;
+}
+
+pub fn decrypt_portion<T: P>(index: T::I) {
+    let iv = S ^ index.into();
+    //~^ ERROR type annotations needed
+    &iv.to_bytes_be();
+}
+
+impl S {
+    fn to_bytes_be(&self) -> &[u8] {
+        &[]
+    }
+}
+
+impl BitXor for S {
+    type Output = S;
+
+    fn bitxor(self, _rhs: Self) -> Self::Output {
+        S
+    }
+}
+
+impl<'a> BitXor<&'a S> for S {
+    type Output = S;
+
+    fn bitxor(self, _rhs: &'a S) -> Self::Output {
+        S
+    }
+}
+
+fn main() {}

--- a/tests/ui/type-inference/regression-issue-81317.rs
+++ b/tests/ui/type-inference/regression-issue-81317.rs
@@ -1,6 +1,7 @@
 // Regression test for #81317: type can no longer be infered as of 1.49
 //
-// The problem is that the xor operator and the index.into() each have two candidate impls that could apply
+// The problem is that the xor operator and the index.into() call
+// each have two candidate impls that could apply
 // { S as BitXor<S>, S as BitXor<&'a S> } for xor and
 // { T::I as Into<u64>, T::I as Into<S> } for index.into()
 // previously inference was able to infer that the only valid combination was
@@ -8,13 +9,14 @@
 //
 // after rust-lang/rust#73905 this is no longer infered
 //
-// the error message could be better e.g. when iv is unused or has an an explicitly specified type S
+// the error message could be better e.g.
+// when iv is unused or has an an explicitly specified type S
 // there is currently the following help message
 //
 // error[E0284]: type annotations needed
 //   --> src/main.rs:13:24
 //    |
-// 42 |     let iv = S ^ index.into();
+// 44 |     let iv = S ^ index.into();
 //    |                -       ^^^^
 //    |                |
 //    |                type must be known at this point
@@ -22,8 +24,8 @@
 //    = note: cannot satisfy `<S as BitXor<_>>::Output == _`
 // help: try using a fully qualified path to specify the expected types
 //    |
-// 42 -     let iv = S ^ index.into();
-// 42 +     let iv = S ^ <<T as P>::I as Into<T>>::into(index);
+// 44 -     let iv = S ^ index.into();
+// 44 +     let iv = S ^ <<T as P>::I as Into<T>>::into(index);
 //
 // this is better as it's actually sufficent to fix the problem,
 // while just specifying the type of iv as currently suggested is insufficent

--- a/tests/ui/type-inference/regression-issue-81317.stderr
+++ b/tests/ui/type-inference/regression-issue-81317.stderr
@@ -1,0 +1,17 @@
+error[E0282]: type annotations needed
+  --> $DIR/regression-issue-81317.rs:13:9
+   |
+LL |     let iv = S ^ index.into();
+   |         ^^
+LL |
+LL |     &iv.to_bytes_be();
+   |      -- type must be known at this point
+   |
+help: consider giving `iv` an explicit type
+   |
+LL |     let iv: /* Type */ = S ^ index.into();
+   |           ++++++++++++
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0282`.

--- a/tests/ui/type-inference/regression-issue-81317.stderr
+++ b/tests/ui/type-inference/regression-issue-81317.stderr
@@ -1,5 +1,5 @@
 error[E0282]: type annotations needed
-  --> $DIR/regression-issue-81317.rs:13:9
+  --> $DIR/regression-issue-81317.rs:44:9
    |
 LL |     let iv = S ^ index.into();
    |         ^^

